keps/sig-network/4410-k8s-network-interface/README.md

```diff
@@ -0,0 +1,113 @@
+markedly different approaches. This fragmentation has spread networking across
+all layers of the stack which include k8s components like kube-proxy, netpol agents,
+container runtime with CNI plugins and low level runtimes like kata and issues
+with the API have negatively impacted adoption in sectors such as telecommunications.
+Our goal is to transform Kubernetes networking by making networks and their components
+actual resources within the Kubernetes API. This will allow for the development
+of shared functionalities and their integration into the API. We anticipate that
+this new approach will enhance support for areas that are currently struggling,
+facilitate the development and promotion of common features, and better define
+and accommodate advanced functionalities and potential areas for expansion.
+
+### Goals
+
+- Design a cool looking t-shirt
+- Provide Kubernetes APIs for the creation, configuration and management of interfaces
+- Provide documentation, examples, troubleshooting and FAQ's for KNI.
+- KNI should provide the API's required to establish feature parity with current CNI [ADD, DEL]
+- Handle support levels like Gateway API (e.g. "core" and "extended")
+- Handle implementation-specific use cases through extension points
+- Decouple the Pod and Node Network setup
+- Provide garbage collection to ensure no resources created during pod setup such as Linux bridges, ebpf programs, 
+allocated IP addresses are left behind after pod deletion
+- Improve the current IP handling for pods (PodIP) to be handle multiple IP addresses and 
+a field to identify the IP address family (IPV4 vs IPV6)
+- Provide backwards compatibility for the existing CNI approach and migration a path to fully adopt KNI
+- Guarantee the network is setup and in a healthy state before containers are started (ephemeral, init, regular)
+- If feasible, provide API awareness of Pod network namespaces (e.g. interface names)
+- Provide a uniform approach for network setup/teardown for both virtualized (kata) and non-virtualized (runc) 
+runtimes including kubevirt. This could eliminate the high and low level runtimes from the networking path
+- Provide a reference implementation of the KNI network runtime
+- Provide the ability to have all the dependencies packaged in the container image (no more CNI binaries in the host file system)
+..- No more downloading CNI binaries via initContainers/Mounting /etc/cni/net.d or /opt/cni/bin
+- Provide the ability to use native k8s resources for configuration such as a ConfigMap's instead of configuration files in host file system
+- Provide an API to indicate network readiness for the node (no more files on disk)
+- Eliminate the need to exec binaries and replace with gRPC
+- Make troubleshooting easier by having logs accessible via kubectl logs
+- Improve network pod startup time
+- Provide the ability to prevent additional scheduling of pods if IPAM is out of IP addresses without evicting running pods
+
```

Comment:
TODO: Add goal of having the pod object available at network runtime

Reply:
@dougbtv I am drafting an update. So I might be able to get this. Do you have specific items you want off the Pod spec? Metadata (name, namespace, labels, annotations, ... )

Reply:
Metadata nails it, thanks. At least I'm most interested in getting all you listed. Potentially someone might want more?

Reply:
I'd like to request CIDRs as an available piece of metadata if possible. That would be great for legacy applications that use CIDR configurations as config values (in my case, Ceph).

Reply:
@BlaineEXE this sounds reasonable. I notice you are in Colorado. I am in the boulder area. Does the application need the pod cidr? You might be able to infer this via the Pod IP.

Reply:
Hey! Yeah, I'm in Denver. And also a climber :)

Our storage application [Ceph](https://ceph.io) does require CIDR in certain circumstances. In "basic" operation, we don't need to provide the CIDR of the pod network.

But when we use Multus to attach pods to host devices (described more [here](https://github.com/kubernetes/enhancements/pull/4477/files#r1486407995)), we need to configure Ceph to listen on a CIDR (or list of CIDRs) so that it knows which interface(s) to use. I don't fully grok why Ceph uses CIDR instead of iface names, but it's what we have to work with.

Currently, what we have been doing is to use the JSON output of `ip addr` in a K8s Job to discover the CIDR of the network. It has been working for us, but it's a step that would be nice to eliminate if possible.

Reply:
Let's climb! Are you at the spot or movement? This sounds like a story for the KNI that could enhance the downward API information with additional node specific network information such as Pod CIDR?

Reply:
> . I don't fully grok why Ceph uses CIDR instead of iface names, but it's what we have to work with.

we need to be careful to not absorb other projects problems, and avoid bespoke solutions

Reply:
@MikeZappa87 I'm with the Spot and usually get punches for mvmt.

@aojea I agree that this project shouldn't absorb other problems. I can offer up our use case (attaching pods to raw devices, potentially with macvlan) as something that seems worthwhile for KNI to consider. 

I personally think that this part of our use case is extremely valuable applications that want to achieve maximum network performance while also benefitting from the superior scheduling offered by Kubernetes. 

I have been conceptualizing this similarly to how Kubernetes allows workflows that integrate GPUs into pod environments. Some nodes have special, high-performance hardware available to provide acceleration for certain workflows. As GPUs are workflow accelerators, so too are networks.

As far as the CIDR goes, my experience has been that every network has one, and I think it is potentially useful for that to be part of the metadata made available to pods. However, if there are deeper complexities with exposing CIDR metadata that I'm not aware of, it won't make or break my project to not have that included in KNI; we are already working around the lack of that information in the existing sig-network APIs.

Reply:
>  I can offer up our use case (attaching pods to raw devices, potentially with macvlan) as something that seems worthwhile for KNI to consider.

that is exactly what we need, these experiences, and this one is something I've identified in multiple places, attach netdevices to pods, so I feel this is a strong use case ... what I also see is that these interfaces are used as "external' networks that are only relevant to the app running on the specific pod, so I don't feel that these IPs from these interfaces should be represented on the kubernetes topology ...

Reply:
@BlaineEXE I'm still trying to fully understand your use case, based on your comments it seems you need to have some prior work of setting up the infrastructure and the vlans,
who configures these vlans?
are these vlans configured on all these hosts?
...

Reply:
Certainly someone must configure the additional hardware. In practice, an admin must add a separate switch (or create a VLAN on an existing switch) that connects to a different interface on the host systems. So if `eth0` underpins the k8s pod network, `eth1` may be the result of the additional interface, unused by k8s itself.

Our current deployment strategy leverages Multus and NetworkAttachmentDefinitions to connect storage (Ceph) pods to `eth1`. We recommend CNI=macvlan (lower latency than bridge) with IPAM=whereabouts (ease of use), but users can practically use whatever CNI/IPAM they like.

This does work, but because Multus is such a complex feature to understand, users often seem lost trying to configure an already-complex storage system with NADs. In addition, there is developer complexity, and there are friction points -- like not being able to get a Service with a static IP on a Multus network.

Reply:
>  like not being able to get a Service with a static IP on a Multus network.

what do you mean by a multus network? some entity that is connected to the additional interface?

Reply:
@BlaineEXE @dougbtv I want to understand well this use case, I don't quite get what are the expectations here and what is the problem we are trying to solve., definitively kubernetes can not manage the infra, connecting switches or create vlanes, there are other projects that cover that area ... 
but this part of _Service with staticIP on Multus network_ is the one I want to understand, the docs are not clear  https://github.com/rook/rook/blob/master/design/ceph/multus-network.md, does it mean to make services available out of the cluster?